### ISSUE TITLE
modifying is_punct lost effect after saving to disk

### ISSUE THREAD

**Issue by xtfocus** [explosion/spaCy#12938]
I'm on spacy 3.5

I need to build an entity ruler that sees `_` as a punct while see `#` as not a punct, here's what I tried:

```import spacy
from spacy.lang.en import English
from spacy.tokenizer import Tokenizer
from spacy.util import compile_infix_regex

def modify_english_model():
    """
    Append an entity_ruler(whose internal is a PhraseMatcher)
        to a pretrained Spacy English model.
    Customize the Tokenizer of the English model to allow splitting on
underscores

    """
    nlp = spacy.load("en_core_web_sm")
    nlp.disable_pipe("ner")
    # Customize the tokenizer
    nlp.tokenizer = custom_tokenizer(nlp)
    nlp.vocab["#"].is_punct = False
    nlp.add_pipe("entity_ruler")
    return nlp

def custom_tokenizer(nlp):
    """
    Customize the out-of-the-box Spacy Tokenizer
        so that it see underscore _ as a punctuation
    """
    infixes = nlp.Defaults.infixes + [
        r"(?<=[0-9a-z])[_](?=[0-9a-z])"
    ]  # Add underscore infix
    infix_re = compile_infix_regex(infixes)

    return Tokenizer(
        nlp.vocab,
        prefix_search=nlp.tokenizer.prefix_search,
        suffix_search=nlp.tokenizer.suffix_search,
        infix_finditer=infix_re.finditer,
        token_match=nlp.tokenizer.token_match,
        rules=nlp.Defaults.tokenizer_exceptions,
    )



model = modify_english_model()

model.to_disk("sample_model")
print(model.vocab['#'].is_punct)

model = spacy.load('sample_model')
print(model.vocab['#'].is_punct)
```

Which prints 
```
False
True
```

It seems that model.vocab['#'].is_punct lost effects after saving the model, which is not what I wanted. How am I supposed to do this? Thank you

**Comment by svlandeg** [comment]
Hi @xtfocus,

Thanks for the detailed code snippet - that always helps to understand what you're trying to accomplish!

Fundamentally, the `is_punct` functionality is a [Lexeme](https://spacy.io/api/lexeme) property, and I don't think it's a good idea to be changing these after a pipeline has already been trained (like `en_core_web_sm` is).

I think perhaps the cleanest solution to do this, is to define a custom language (sub)class that overrides the `IS_PUNCT` functionality in `lex_attr_getters`, originally defined here: https://github.com/explosion/spaCy/blob/master/spacy/lang/lex_attrs.py#L25

So you could do something like this:
```
from spacy.attrs import IS_PUNCT
from spacy.lang.en import English, EnglishDefaults, LEX_ATTRS


def custom_punct(text: str) -> bool:
    for char in text:
        if char == "#":
            return False
        if not unicodedata.category(char).startswith("P"):
            return False
    return True


class CustomEnglishDefaults(EnglishDefaults):
    lex_attr_getters = LEX_ATTRS
    lex_attr_getters[IS_PUNCT] = custom_punct


@spacy.registry.languages("custom_en")
class CustomEnglish(English):
    lang = "custom_en"
    Defaults = CustomEnglishDefaults
```
And then if you create a new pipeline with `spacy.blank("custom_en")`, the `is_punct` of `"#"` will be `False`.

**Comment by xtfocus** [comment]
> Hi @xtfocus,
> 
> Thanks for the detailed code snippet - that always helps to understand what you're trying to accomplish!
> 
> Fundamentally, the `is_punct` functionality is a [Lexeme](https://spacy.io/api/lexeme) property, and I don't think it's a good idea to be changing these after a pipeline has already been trained (like `en_core_web_sm` is).
> 
> I think perhaps the cleanest solution to do this, is to define a custom language (sub)class that overrides the `IS_PUNCT` functionality in `lex_attr_getters`, originally defined here: https://github.com/explosion/spaCy/blob/master/spacy/lang/lex_attrs.py#L25
> 
> So you could do something like this:
> 
> ```
> from spacy.attrs import IS_PUNCT
> from spacy.lang.en import English, EnglishDefaults, LEX_ATTRS
> 
> 
> def custom_punct(text: str) -> bool:
>     for char in text:
>         if char == "#":
>             return False
>         if not unicodedata.category(char).startswith("P"):
>             return False
>     return True
> 
> 
> class CustomEnglishDefaults(EnglishDefaults):
>     lex_attr_getters = LEX_ATTRS
>     lex_attr_getters[IS_PUNCT] = custom_punct
> 
> 
> @spacy.registry.languages("custom_en")
> class CustomEnglish(English):
>     lang = "custom_en"
>     Defaults = CustomEnglishDefaults
> ```
> 
> And then if you create a new pipeline with `spacy.blank("custom_en")`, the `is_punct` of `"#"` will be `False`.

Thanks. I used the pre-trained model because I wanted to be able to detect entities like organizations (universities, etc.). I there a way to add that ner component to the blank model?

**Comment by svlandeg** [comment]
Yes, exactly. And in this particular case it should hopefully work just fine because AFAIK the pretrained NER model doesn't use the `is_punct` information from lexemes. But it's the kind of thing where you'll want to benchmark on your data that the solution works with sufficient accuracy.

I'll go ahead and close this issue ;-)
